pkg/actuators/machine/conditions_test.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machine
+
+import (
+	"testing"
+
+	"github.com/openshift/cluster-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestShouldUpdateCondition(t *testing.T) {
+	testCases := []struct {
+		oldCondition v1beta1.IBMCloudMachineProviderCondition
+		newCondition v1beta1.IBMCloudMachineProviderCondition
+		expected     corev1.ConditionStatus
+	}{
+		{
+			oldCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionTrue,
+			},
+			newCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionTrue,
+			},
+		},
+		{
+			oldCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionTrue,
+			},
+			newCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo2",
+				Message: "bar2",
+				Status:  corev1.ConditionTrue,
+			},
+		},
+		{
+			oldCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionTrue,
+			},
+			newCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo2",
+				Message: "New Message",
+				Status:  corev1.ConditionFalse,
+			},
+		},
+		{
+			oldCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionTrue,
+			},
+			newCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "New Reason",
+				Message: "New Message",
+				Status:  corev1.ConditionTrue,
+			},
+		},
+		{
+			oldCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionFalse,
+			},
+			newCondition: v1beta1.IBMCloudMachineProviderCondition{
+				Reason:  "foo",
+				Message: "bar",
+				Status:  corev1.ConditionTrue,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+
+		conditions := []v1beta1.IBMCloudMachineProviderCondition{}
+		conditions = append(conditions, tc.oldCondition)
+		returnCondition := reconcileProviderConditions(conditions, tc.newCondition)
+
+		if returnCondition[0].Reason != tc.newCondition.Reason &&
```

Comment:
Just eyeballing this one @gsrajpal ... shouldn't the first test case element _(lines 33-42)_ fail this condition?
The `reason`, `message`, and `status` fields appear to be the same between old and new conditions in that first element whereas the other test case elements differ in some way.

Reply:
So the first test mimics a scenario where the Current Conditions and New Conditions are the same - the expected behavior should be `No change` -- essentially just checking if the conditions are updated to the NewCondtions every time.

Reply:
Ok - I've had a bit more coffee now 😄 
So, the check here is that the original [0] condition should match the new one, otherwise there's an error?
I would think that would be true then for the first test case element (no error, no difference), but not for the latter ones whose conditions differ between old vs new - i.e. returnCondition[0] would contain the old value (!= new), but returnCondition[1] would equate to the new condition set.

Reply:
`So, the check here is that the original [0] condition should match the new one, otherwise there's an error?`
- _Yes_

`I would think that would be true then for the first test case element (no error, no difference), but not for the latter ones whose.....`
- _Think of each item in the struct as a separate test case. Every time a test is executed, there is only one item in `conditions` - the `latter` test cases `oldConditon & newCondition` have the same Condition `Type` which executes the logic to update the currCondition rather than `appending` a new object. So `returnCondition` only has one item as well, then essentially  we just compare `returnContion[0] & newCondition`_

Reply:
I get it now ... conditions of the same type are updated rather than appended.
Couple additional things you could check here : 
1. Make sure `returnCondition` only contains a single instance of the condition passed in vs being appended. I believe this is being implicitly tested in that test cases where old vs new would fail. It's not something we check for in the `conditionTypeCheck` function.
2. Check that the last transition time was set when the status value differed and maybe the last probe times as well.
3. Pass in conditions of different types, make sure the correct condition was updated. We're only passing in 1 old and 1 new of the same time at the moment.
4. Pass in an old set of conditions that do not contain the instance type of the new condition - make sure it gets added.

Reply:
1. Technically we should always have one object since there is only one Condition type `MachineCreated`.
   - Conditions will either be empty or it will only have `MachineCreated` type.
   - If the condition is empty (the type was not found), we will append a new condition (this should only happen once - making the `conditions[]` `len() = 1`) otherwise check if any of the fields are different and update the condition object accordingly. 
2. This is in a way covered as well, but we are not explicitly checking. If the `Status` is different, the `LastTransitionTime` is updated. 
   - We know this must be happening because we are making sure the other fields match the expected `newCondition` (`LastTransitionTime` update is happening at the same level)
   - Maybe I can add an explicit test case for this. 
3. We should technically only have one Type: `MachineCreated`
4. Not sure when this can occur. If the Condition Type is missing from the old then we will surely append a new object. I guess we could add this test, but not sure if this is a valid test. 

Do you foresee us having any other condition types?

Reply:
good point - since the conditions are handled within an array, I had presumed other condition types might exist, maybe `MachineFailure` or something
I presume then that we're using an array to accommodate other condition types in the future, even though there's only one at the moment?
Didn't look as though Azure or GCP had condition types other than `MachineCreated` either.

Reply:
AWS and GCP have only one condition as well.